Incorrect mscorlib selected by msbuild

When generating the `.csproj` files, Rider 2018.3 will add a hint path to all explicitly referenced system assemblies to point to the current Unity version's Mono install. This usually affects `System`, `System.Xml`, `System.Core`, `System.Runtime.Serialization` and `System.Xml.Linq`.

However, msbuild also adds an explicit reference to `mscorlib`, as well as a secondary reference to `System.Core`. These references have no hint path and are resolved by the target framework version.

[Unity 2017.2](https://github.com/Unity-Technologies/UnityCsReference/commit/9be0e8275aa1aa1ece02d2284bca53c2956699a0#diff-ae1ccef0c7a021d0bcea8ed33e811075R376) to (but not including) [Unity 2018.3.0a9](https://github.com/Unity-Technologies/UnityCsReference/commit/555f53cb7e0b443f6db8fc227f062c73f5b96990#diff-ae1ccef0c7a021d0bcea8ed33e811075L494) (links take a while to fully load) would hard code the target framework version as `v4.5` for Rider, even for a project targeting .NET 2.0/3.5.

This means that we can get into the situation where `System`, `System.Xml`, etc. are pointing at Unity's `2.0-api` Mono folder, and msbuild resolves `mscorlib` against the system installed .NET 4.5. The secondary reference of `System.Core` is auto-unified to the resolved version - 4.5 in the system installed .NET folder.

This can cause compile errors, such as:

```
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/TestRunner.cs(134,13,134,28): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/ConsolePro/Remote/LiteNetLib/NetSerializer.cs(118,9,131,10): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/IntegrationTestsProvider.cs(43,24,43,62): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/IntegrationTestsProvider.cs(67,17,67,46): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/UnityTestTools/IntegrationTestsFramework/TestRunner/IntegrationTestsProvider.cs(104,17,104,40): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
Assets/Plugins/Trinary Software/Timing.cs(1477,69,1477,80): error CS7069: Reference to type 'ISet<>' claims it is defined in 'System', but it could not be found
```

(Note that I haven't been able to recreate this error locally. I get unresolved symbols for `ISet<>` in ReSharper and when building. I think you might need to include a reference to something that causes one of the system installed .NET system libraries to be referenced.)

I think this will fix it:

* Set `<NoStdLib>` to `true`. This will tell msbuild to not add the explicit reference to `mscorlib`
* Set `<AddAdditionalExplicitAssemblyReferences>` to `false`. This stops msbuild adding an extra reference to `System.Core` without a hint path, which then gets resolved to the wrong version and auto-unified, overriding the explicit reference in the project
* Add an explicit reference to `mscorlib`, with a hint path that matches `System`

Optionally, we can also do one or more of the following:

* Set `<AssemblySearchPaths>` to `{HintPathFromItem}`. This clears all search paths, and requires that references must have a hint path. Alternatively, we could copy most of these items, but remove `{TargetFrameworkDirectory}` and the `{Registry:…}` entries so that we'll search in normal places, but never resolve against the target framework directory.

    ```
    <!--
        The SearchPaths property is set to find assemblies in the following order:

            (1) Files from current project - indicated by {CandidateAssemblyFiles}
            (2) $(ReferencePath) - the reference path property, which comes from the .USER file.
            (3) The hintpath from the referenced item itself, indicated by {HintPathFromItem}.
            (4) The directory of MSBuild's "target" runtime from GetFrameworkPath.
                The "target" runtime folder is the folder of the runtime that MSBuild is a part of.
            (5) Registered assembly folders, indicated by {Registry:*,*,*}
            (6) Assembly folders from AssemblyFolders.config file (provided by Visual Studio Dev15+).
            (7) Legacy registered assembly folders, indicated by {AssemblyFolders}
            (8) Resolve to the GAC.
            (9) Treat the reference's Include as if it were a real file name.
            (10) Look in the application's output folder (like bin\debug)
        -->
    <AssemblySearchPaths Condition=" '$(AssemblySearchPaths)' == ''">
      {CandidateAssemblyFiles};
      $(ReferencePath);
      {HintPathFromItem};
      {TargetFrameworkDirectory};
      $(AssemblyFoldersConfigFileSearchPath)
      {Registry:$(FrameworkRegistryBase),$(TargetFrameworkVersion),$(AssemblyFoldersSuffix)$(AssemblyFoldersExConditions)};
      {AssemblyFolders};
      {GAC};
      {RawFileName};
      $(OutDir)
    </AssemblySearchPaths>
    ```

* Set the target framework directory to the correct folder. We could probably leave the system assemblies without hint paths and let msbuild find them in the new target framework dir. This would mean revisiting `AvoidGetReferenceAssemblyCalls`. We should probably look at this for pre-Unity 2018.1 too.

--- a/rider/build.gradle
+++ b/rider/build.gradle
@@ -5,7 +5,7 @@ plugins {
     id 'java'
     id 'org.jetbrains.kotlin.jvm' version '1.3.0'
     id 'org.jetbrains.intellij' // Version comes from buildSrc
-    id 'com.jetbrains.rdgen' version '0.183.86'
+    id 'com.jetbrains.rdgen' version '0.183.101'
     id 'com.ullink.nuget' version '2.17'
     id 'com.ullink.nunit' version '1.12'
 }

--- a/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
+++ b/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
@@ -122,11 +122,11 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
       changed |= SetManuallyDefinedCompilerSettings(projectFile, projectContentElement, xmlns);
       changed |= TrySetHintPathsForSystemAssemblies(projectContentElement, xmlns);
       changed |= FixImplicitReferences(projectContentElement, xmlns);
+      changed |= AvoidGetReferenceAssemblyPathsCall(projectContentElement, xmlns);
       changed |= AddMicrosoftCSharpReference(projectContentElement, xmlns);
       changed |= SetXCodeDllReference("UnityEditor.iOS.Extensions.Xcode.dll", projectContentElement, xmlns);
       changed |= SetXCodeDllReference("UnityEditor.iOS.Extensions.Common.dll", projectContentElement, xmlns);
       changed |= SetDisableHandlePackageFileConflicts(projectContentElement, xmlns);
-      changed |= AvoidGetReferenceAssemblyPathsCall(projectContentElement, xmlns);
       changed |= SetGenerateTargetFrameworkAttribute(projectContentElement, xmlns);
       
       return changed;
@@ -145,18 +145,29 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
     // https://github.com/JetBrains/resharper-unity/issues/988
     private static bool FixImplicitReferences(XElement projectContentElement, XNamespace xmlns)
     {
-      // Starting with Unity 2018.2, it must be done by Unity itself
-      if (UnityUtils.UnityVersion >= new Version(2018, 2))
-        return false;
-      
-      var changed = SetOrUpdateProperty(projectContentElement, xmlns, "NoConfig", existing => "true");
-      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "NoStdLib", existing => "true");
-      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "AddAdditionalExplicitAssemblyReferences", existing => "false");
-      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandNETStandardFacades", existing => "false");
-      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandDesignTimeFacades", existing => "false");
+      var changed = false;
+
+      // For Unity 2017.x, we are adding tags and reference to mscorlib
+      if (UnityUtils.UnityVersion.Major == 2017)
+      {
+        changed = SetOrUpdateProperty(projectContentElement, xmlns, "NoConfig", existing => "true");
+        changed |= SetOrUpdateProperty(projectContentElement, xmlns, "NoStdLib", existing => "true");
+        changed |= SetOrUpdateProperty(projectContentElement, xmlns, "AddAdditionalExplicitAssemblyReferences",existing => "false");
+
+        var referenceName = "mscorlib.dll";
+        var hintPath = GetHintPath(referenceName);
+        AddCustomReference(referenceName, projectContentElement, xmlns, hintPath);
+      }
+
+      if (UnityUtils.UnityVersion.Major == 2017 || UnityUtils.UnityVersion.Major == 2018)
+      {
+        changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandNETStandardFacades",existing => "false");
+        changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandDesignTimeFacades",existing => "false");
+      }
+
       return changed;
     }
-    
+
     // Computer may not have specific TargetFramework, msbuild will resolve System from different TargetFramework
     // If we set HintPaths together with DisableHandlePackageFileConflicts we help msbuild to resolve libs from Unity installation
     // Unity 2018+ already have HintPaths by default
@@ -210,8 +221,8 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
     
     private static bool AvoidGetReferenceAssemblyPathsCall(XElement projectContentElement, XNamespace xmlns)
     {
-      // Starting with Unity 2018, dotnet target pack is not required
-      if (UnityUtils.UnityVersion < new Version(2018, 1))
+      // Starting with Unity 2017, dotnet target pack is not required
+      if (UnityUtils.UnityVersion.Major < 2017)
         return false;
       
       // Set _TargetFrameworkDirectories and _FullFrameworkReferenceAssemblyPaths to something to avoid GetReferenceAssemblyPaths task being called

--- a/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
+++ b/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
@@ -132,8 +132,7 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
       return changed;
     }
     
-    /* In Unity 2018.2 it looks like this:
-     
+    /* Since Unity 2018.1.5f1 it looks like this:
      <PropertyGroup>
            <NoConfig>true</NoConfig>
            <NoStdLib>true</NoStdLib>

--- a/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
+++ b/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
@@ -121,6 +121,7 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
       changed |= SetProjectFlavour(projectContentElement, xmlns);
       changed |= SetManuallyDefinedCompilerSettings(projectFile, projectContentElement, xmlns);
       changed |= TrySetHintPathsForSystemAssemblies(projectContentElement, xmlns);
+      changed |= FixImplicitReferences(projectContentElement, xmlns);
       changed |= AddMicrosoftCSharpReference(projectContentElement, xmlns);
       changed |= SetXCodeDllReference("UnityEditor.iOS.Extensions.Xcode.dll", projectContentElement, xmlns);
       changed |= SetXCodeDllReference("UnityEditor.iOS.Extensions.Common.dll", projectContentElement, xmlns);
@@ -128,6 +129,31 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
       changed |= AvoidGetReferenceAssemblyPathsCall(projectContentElement, xmlns);
       changed |= SetGenerateTargetFrameworkAttribute(projectContentElement, xmlns);
       
+      return changed;
+    }
+    
+    /* In Unity 2018.2 it looks like this:
+     
+     <PropertyGroup>
+           <NoConfig>true</NoConfig>
+           <NoStdLib>true</NoStdLib>
+           <AddAdditionalExplicitAssemblyReferences>false</AddAdditionalExplicitAssemblyReferences>
+           <ImplicitlyExpandNETStandardFacades>false</ImplicitlyExpandNETStandardFacades>
+           <ImplicitlyExpandDesignTimeFacades>false</ImplicitlyExpandDesignTimeFacades>
+         </PropertyGroup>
+    */
+    // https://github.com/JetBrains/resharper-unity/issues/988
+    private static bool FixImplicitReferences(XElement projectContentElement, XNamespace xmlns)
+    {
+      // Starting with Unity 2018.2, it must be done by Unity itself
+      if (UnityUtils.UnityVersion < new Version(2018, 2))
+        return false;
+      
+      var changed = SetOrUpdateProperty(projectContentElement, xmlns, "NoConfig", existing => "true");
+      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "NoStdLib", existing => "true");
+      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "AddAdditionalExplicitAssemblyReferences", existing => "false");
+      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandNETStandardFacades", existing => "false");
+      changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandDesignTimeFacades", existing => "false");
       return changed;
     }
     

--- a/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
+++ b/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
@@ -150,13 +150,14 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
       // For Unity 2017.x, we are adding tags and reference to mscorlib
       if (UnityUtils.UnityVersion.Major == 2017)
       {
-        changed = SetOrUpdateProperty(projectContentElement, xmlns, "NoConfig", existing => "true");
-        changed |= SetOrUpdateProperty(projectContentElement, xmlns, "NoStdLib", existing => "true");
-        changed |= SetOrUpdateProperty(projectContentElement, xmlns, "AddAdditionalExplicitAssemblyReferences",existing => "false");
+        SetOrUpdateProperty(projectContentElement, xmlns, "NoConfig", existing => "true");
+        SetOrUpdateProperty(projectContentElement, xmlns, "NoStdLib", existing => "true");
+        SetOrUpdateProperty(projectContentElement, xmlns, "AddAdditionalExplicitAssemblyReferences",existing => "false");
 
         var referenceName = "mscorlib.dll";
         var hintPath = GetHintPath(referenceName);
         AddCustomReference(referenceName, projectContentElement, xmlns, hintPath);
+        changed = true;
       }
 
       if (UnityUtils.UnityVersion.Major == 2017 || UnityUtils.UnityVersion.Major == 2018)

--- a/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
+++ b/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
@@ -150,6 +150,7 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
       // For Unity 2017.x, we are adding tags and reference to mscorlib
       if (UnityUtils.UnityVersion.Major == 2017)
       {
+        // appears in Unity 2018.1.0b10
         SetOrUpdateProperty(projectContentElement, xmlns, "NoConfig", existing => "true");
         SetOrUpdateProperty(projectContentElement, xmlns, "NoStdLib", existing => "true");
         SetOrUpdateProperty(projectContentElement, xmlns, "AddAdditionalExplicitAssemblyReferences",existing => "false");
@@ -163,6 +164,7 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
 
       if (UnityUtils.UnityVersion.Major == 2017 || UnityUtils.UnityVersion.Major == 2018)
       {
+        // appears in Unity 2018.1.5f1
         changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandNETStandardFacades",existing => "false");
         changed |= SetOrUpdateProperty(projectContentElement, xmlns, "ImplicitlyExpandDesignTimeFacades",existing => "false");
       }

--- a/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
+++ b/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
@@ -146,7 +146,7 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
     private static bool FixImplicitReferences(XElement projectContentElement, XNamespace xmlns)
     {
       // Starting with Unity 2018.2, it must be done by Unity itself
-      if (UnityUtils.UnityVersion < new Version(2018, 2))
+      if (UnityUtils.UnityVersion >= new Version(2018, 2))
         return false;
       
       var changed = SetOrUpdateProperty(projectContentElement, xmlns, "NoConfig", existing => "true");

--- a/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
+++ b/unity/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs
@@ -154,6 +154,7 @@ namespace JetBrains.Rider.Unity.Editor.AssetPostprocessors
         SetOrUpdateProperty(projectContentElement, xmlns, "NoStdLib", existing => "true");
         SetOrUpdateProperty(projectContentElement, xmlns, "AddAdditionalExplicitAssemblyReferences",existing => "false");
 
+        // Unity 2018.x+ itself adds mscorlib reference 
         var referenceName = "mscorlib.dll";
         var hintPath = GetHintPath(referenceName);
         AddCustomReference(referenceName, projectContentElement, xmlns, hintPath);

